core/src/main/java/org/ehcache/events/CacheEventNotificationServiceImpl.java

```diff
@@ -170,64 +264,61 @@ public void run() {
-
-  private static final class EventListenerWrapper {
-    final CacheEventListener<?, ?> listener;
-    final CacheEventListenerConfiguration config;
-    
-    EventListenerWrapper(CacheEventListener<?, ?> listener, final EventFiring firing, final EventOrdering ordering, 
-        final EnumSet<EventType> forEvents) {
-      this.listener = listener;
-      this.config = new CacheEventListenerConfiguration() {
-        
-        @Override
-        public Class<CacheEventListenerProvider> getServiceType() {
-          return CacheEventListenerProvider.class;
-        }
-        
-        @Override
-        public EventOrdering orderingMode() {
-          return ordering;
-        }
-        
-        @Override
-        public EventFiring firingMode() {
-          return firing;
-        }
-        
-        @Override
-        public EnumSet<EventType> fireOn() {
-          return forEvents;
+    return !syncListenersSet.isEmpty() || !aSyncListenersSet.isEmpty();
+  }
+
+  @Override
+  public List<CacheConfigurationChangeListener> getConfigurationChangeListeners() {
+    List<CacheConfigurationChangeListener> configurationChangeListenerList = new ArrayList<CacheConfigurationChangeListener>();
+    configurationChangeListenerList.add(new CacheConfigurationChangeListener() {
+      @Override
+      public void cacheConfigurationChange(final CacheConfigurationChangeEvent event) {
+        if (event.getProperty().equals(CacheConfigurationProperty.ADDLISTENER)) {
+          EventListenerWrapper newListener = (EventListenerWrapper)event.getNewValue();
+          CacheEventListenerConfiguration listenerConfiguration = newListener.config;
+          registerCacheEventListener(newListener.getListener(), listenerConfiguration.orderingMode(),
+              listenerConfiguration.firingMode(), listenerConfiguration.fireOn());
+        } else if (event.getProperty().equals(CacheConfigurationProperty.REMOVELISTENER)) {
+          CacheEventListener<K, V> oldListener = (CacheEventListener)event.getOldValue();
+          deregisterCacheEventListener(oldListener);
         }
-      };
-    }
-    
-    @SuppressWarnings("unchecked")
-    <K, V> CacheEventListener<K, V> getListener() {
-      return (CacheEventListener<K, V>) listener;
-    }
+      }
+    });
+    return configurationChangeListenerList;
+  }
 
-    @Override
-    public int hashCode() {
-      return listener.hashCode();
+  @Override
+  public void fireAllEvents() {
+    EventThreadLocal currentThreadLocal = eventThreadLocal;
+    while (!currentThreadLocal.get().isEmpty()) {
+      CacheEventWrapper cacheEventWrapper = currentThreadLocal.get().get(0);
+      dispatchEvent(cacheEventWrapper, currentThreadLocal);
+      currentThreadLocal.get().remove(0);
     }
-    
-    @Override
-    public boolean equals(Object other) {
-      if (!(other instanceof EventListenerWrapper)) {
-        return false;
+  }
+
+  @Override
+  public void processAndFireRemainingEvents() {
```

Comment:
Objective – To be used when the operation failed and only the associated eviction and expiration events get fired.

What it internally does:
•	Mark all cache events (such as create, update, remove) as failed
•	Call fireAllEvents()
•	Finally cleanUp the threadLocal

Why it should work:
This is called in case the operation has failed. The the threadLocal eventList contains all the events created during the operation so all operations that are not marked as failed, will be eligible for dispatch using the fireAllEvents() itself.